### PR TITLE
Resolve VCSWP-19101

### PR DESCRIPTION
**Pull request by amahadaya**
Closes VCSWP-19101

### CODE REVIEW

File: lib/freeclimb/utils/request_verifier.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        end
+
+        def checkSigningSecret(signingSecret)
+            if signingSecret == "" or signingSecret == nil
+                raise 'Signing secret cannot be empty or null'
+            end
+        end
+
+        def checkTolerance(tolerance)
+            fixnum = (2**(0.size * 8 -2) -1)
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb]: This is not an accurate way to check if the number is exceeding the max integer. Running this locally and I can increase the value of it to a much larger value than whatever the result of this is.

File: lib/freeclimb/utils/request_verifier.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            if tolerance <= 0 or tolerance == Float::NAN or tolerance >= fixnum
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb]: This is incorrect use of ruby `or`. http://phrogz.net/programmingruby/language.html#table_18.4 
The `or` `and` `not` keywords are bad smells with ruby. 
```
if tolerance <= 0 || !(tolerance.is_a? Integer)
```
is a more appropriate conditional

File: lib/freeclimb/utils/request_verifier.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            end
+        end
+
+        def verifySignature(info, requestBody, signingSecret)
+            if !info.isSignatureSafe(requestBody, signingSecret)
+                raise "Unverified signature request, If this request was unexpected, it may be from a bad actor. Please proceed with caution. If the request was exepected, please check any typos or issues with the signingSecret"
+            end
+        end
+    end
+end
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb:61]: Ruby conventions issues. this needs to be applied everywhere.
1. snake case all method & variable names
2. Avoid all usage of `or`, `and`, `not` - only use traditional comparison operators

File: lib/freeclimb/utils/signature_information.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        attr_accessor :requestTimestamp
+        attr_accessor :signatures
+
+        def initialize(requestHeader)
+            @requestTimestamp = 0
+            @signatures = []
+            signatureHeader = requestHeader.try(:split, ",")
+            signatureHeader.each { |signature|
+                header = signature.try(:split, "=").at(0)
+                value = signature.try(:split, "=").at(1)
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/signature_information.rb]: why do multiple splits?
you can easily destructuring instead. 
```
header, value = signature.try(:split, "=")
```

File: spec/utils/signature_information_spec.rb [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+        end
+    end
+    describe '#isSignatureSafe' do
+        context 'signingSecret exists in signature array' do
+            it 'returns true' do
+                requestBody = "{\"accountId\":\"AC1334ffb694cd8d969f51cddf5f7c9b478546d50c\",\"callId\":\"CAccb0b00506553cda09b51c5477f672a49e0b2213\",\"callStatus\":\"ringing\",\"conferenceId\":null,\"direction\":\"inbound\",\"from\":\"+13121000109\",\"parentCallId\":null,\"queueId\":null,\"requestType\":\"inboundCall\",\"to\":\"+13121000096\"}"
+                signingSecret = "sigsec_ead6d3b6904196c60835d039e91b3341c77a7793"
+                expect(@signature_information_object.isSignatureSafe(requestBody, signingSecret)).to be true
+            end
+            it 'returns false' do
```
Comment (ngarg-vail) [inline spec/utils/signature_information_spec.rb]: this should be part of a separate context. why would it return false if the signature is part of the signature array?

**Review by ngarg-vail — CHANGES_REQUESTED** [review]

File: lib/freeclimb/utils/request_verifier.rb (hunk 1 above)
Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb]: So from what I've been researching, it seems that integers don't have a maximum value in Ruby: https://stackoverflow.com/a/47894026

So what I'm thinking is creating a class for the maximum integer so then we can use it within the two util files

I can use this as reference and create an class for this:
https://gist.github.com/pithyless/9738125

Will add this in latest commit

Comment (ngarg-vail) [reply lib/freeclimb/utils/request_verifier.rb]: If there is no limit, then there is no reason to validate against it. Simply checking that it is a positive integer is enough

Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb]: will change the conditional in latest commit

File: lib/freeclimb/utils/request_verifier.rb (hunk 3 above)
Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb:61]: will change in latest commit

File: lib/freeclimb/utils/request_verifier.rb (hunk 2 above)
Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb]: will change in latest commit

File: lib/freeclimb/utils/signature_information.rb (hunk 4 above)
Comment (amahadaya) [reply lib/freeclimb/utils/signature_information.rb]: will change in latest commit

File: spec/utils/signature_information_spec.rb (hunk 5 above)
Comment (amahadaya) [reply spec/utils/signature_information_spec.rb]: will change in latest commit

File: lib/freeclimb/utils/request_verifier.rb [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        def check_request_body(request_body)
+            if request_body == "" || request_body == nil
+                raise 'Request Body cannot be empty or null'
+            end
+        end
+
+        def check_request_header(request_header)
+            if request_header == "" || request_header == nil
+                raise 'Error with request header, Request header is empty'
+            elsif ! request_header.include? "t"
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb]: formatting

File: lib/freeclimb/utils/request_verifier.rb [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+                raise 'Error with request header, timestamp is not present'
+            elsif ! request_header.include? "v1"
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb]: formatting

File: spec/utils/request_verifier_spec.rb [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                request_header = "t=1679944186,v1=c3957749baf61df4b1506802579cc69a74c77a1ae21447b930e5a704f9ec4120,v1=1ba18712726898fbbe48cd862dd096a709f7ad761a5bab14bda9ac24d963a6a8"
+                signing_secret = "sigsec_ead6d3b6904196c60835d039e91b3341c77a7793"
+                tolerance = (5 * 60)
+                request_body =  "{\"accountId\":\"AC1334ffb694cd8d969f51cddf5f7c9b478546d50c\",\"callId\":\"CAccb0b00506553cda09b51c5477f672a49e0b2213\",\"callStatus\":\"ringing\",\"conferenceId\":null,\"direction\":\"inbound\",\"from\":\"+13121000109\",\"parentCallId\":null,\"queueId\":null,\"requestType\":\"inboundCall\",\"to\":\"+13121000096\"}"
+                expect { @request_verifier_object.verify_request_signature(request_body, request_header, signing_secret, tolerance) }.not_to raise_error
+            end
+        end
+    end
+    
+end
```
Comment (ngarg-vail) [inline spec/utils/request_verifier_spec.rb:148]: rename all methods in descriptions to correspond with name updates

File: spec/utils/signature_information_spec.rb [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        end
+        context 'signingSecret does not exists in signature array' do
+            it 'returns false since it does not match condition of signingSecret being within signature array' do
+                requestBody = "{\"accountId\":\"AC1334ffb694cd8d969f51cddf5f7c9b478546d50c\",\"callId\":\"CAccb0b00506553cda09b51c5477f672a49e0b2213\",\"callStatus\":\"ringing\",\"conferenceId\":null,\"direction\":\"inbound\",\"from\":\"+13121000109\",\"parentCallId\":null,\"queueId\":null,\"requestType\":\"inboundCall\",\"to\":\"+13121000096\"}"
+                signingSecret = "sigsec_ead6d3b6904196c60835d039e91b3341c77a7794"
+                expect(@signature_information_object.is_signature_safe(requestBody, signingSecret)).to be false
+            end
+        end
+    end
+end
```
Comment (ngarg-vail) [inline spec/utils/signature_information_spec.rb:38]: Rename all methods in descriptions

File: lib/freeclimb/utils/request_verifier.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        end
+
+        def check_signing_secret(signing_secret)
+            if signing_secret == "" || signing_secret == nil
+                raise 'Signing secret cannot be empty or null'
+            end
+        end
+
+        def check_tolerance(tolerance)
+            if tolerance <= 0
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb]: See comment: https://github.com/FreeClimbAPI/ruby-sdk/pull/35#discussion_r1153862008

File: lib/freeclimb/utils/signature_information.rb [hunk 11]
```diff
@@ -0,0 +1,44 @@
+require 'date'
+require 'openssl'
+require 'cgi'
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/signature_information.rb]: what is this dependency for?

File: lib/freeclimb/utils/request_verifier.rb [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    class RequestVerifier
+        @@DEFAULT_TOLERANCE = 5 * 60 * 1000
+        attr_accessor :info
+
+        def verify_request_signature(request_body, request_header, signing_secret, tolerance=DEFAULT_TOLERANCE)
+            check_request_body(request_body)
+            check_request_header(request_header)
+            check_signing_secret(signing_secret)
+            check_tolerance(tolerance)
+            @info = Freeclimb::SignatureInformation.new(request_header)
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb]: Does not need to be an instance variable.

File: lib/freeclimb/utils/request_verifier.rb [hunk 13]
```diff
@@ -0,0 +1,57 @@
+module Freeclimb
+    class RequestVerifier
+        @@DEFAULT_TOLERANCE = 5 * 60 * 1000
+        attr_accessor :info
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb]: Not needed. Specification does not define this attribute is required

**Review by ngarg-vail — CHANGES_REQUESTED** [review]

File: lib/freeclimb/utils/signature_information.rb (hunk 11 above)
Comment (amahadaya) [reply lib/freeclimb/utils/signature_information.rb]: It's not needed, will remove in latest commit

File: lib/freeclimb/utils/request_verifier.rb (hunk 6 above)
Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb]: This will be changed in latest commit

File: lib/freeclimb/utils/request_verifier.rb (hunk 7 above)
Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb]: This will be changed in latest commit

File: spec/utils/request_verifier_spec.rb (hunk 8 above)
Comment (amahadaya) [reply spec/utils/request_verifier_spec.rb:148]: This will be changed in latest commit

File: spec/utils/signature_information_spec.rb (hunk 9 above)
Comment (amahadaya) [reply spec/utils/signature_information_spec.rb:38]: This will be changed in latest commit

File: lib/freeclimb/utils/request_verifier.rb (hunk 10 above)
Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb]: This will be changed in latest commit

File: lib/freeclimb/utils/request_verifier.rb (hunk 12 above)
Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb]: will remove in latest commit

File: lib/freeclimb/utils/request_verifier.rb (hunk 13 above)
Comment (amahadaya) [reply lib/freeclimb/utils/request_verifier.rb]: will remove in latest commit

File: lib/freeclimb/utils/request_verifier.rb [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+            end
+        end
+
+        def verify_signature(info, request_body, signing_secret)
+            if !info.is_signature_safe(request_body, signing_secret)
+                raise "Unverified signature request, If this request was unexpected, it may be from a bad actor. Please proceed with caution. If the request was exepected, please check any typos or issues with the signingSecret"
+            end
+        end
+    end
+end
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb:61]: does not conform to the public/private access defined in the spec

**Review by ngarg-vail — CHANGES_REQUESTED** [review]

File: lib/freeclimb/utils/request_verifier.rb [hunk 15]
```diff
@@ -0,0 +1,61 @@
+module Freeclimb
+    class RequestVerifier
+        class << self
+            @@DEFAULT_TOLERANCE = 5 * 60 * 1000
+            
+            public def verify_request_signature(request_body, request_header, signing_secret, tolerance=DEFAULT_TOLERANCE)
```
Comment (ngarg-vail) [inline lib/freeclimb/utils/request_verifier.rb]: can remove. ruby methods are public by default

**Review by ngarg-vail — CHANGES_REQUESTED** [review]

**Review by ngarg-vail — APPROVED** [review]

**Pull request merged**